File: Sum21.py
class Node:
    def __init__(self):
        self.children = []
        self.moves = []
        self.score_of_move = []


def minimax(node, depth, sum):
    if game_over(sum):
        score = get_score(sum, depth)
        node.score_of_move.append(score)
        node.moves.clear()
        return score

    score = 0

    for move in node.moves:
        child = Node()
        mvs = [moves for moves in node.moves if moves is not move]
        child.moves.extend(mvs)
        node.children.append(child)
        score = minimax(child, depth + 1, sum + move)
        node.score_of_move.append(apt_score(depth, child))

    return score


def game_over(sum):
    return sum >= 21


def get_score(sum, depth):
    if sum > 21:
        if depth % 2 is not 0:
            return 10 - depth
        else:
            return depth - 10
    else:
        return 0


def apt_score(depth, node):
    if depth%2 is not 0:
        return min(node.score_of_move)
    return max(node.score_of_move)


def ai_move(node):
    min_index = -1
    min = 100
    global depth, root
    depth = depth + 1
    for i in range(0, len(node.moves)):
        if node.score_of_move[i] < min:
            min = node.score_of_move[i]
            min_index = i
    root = node.children[min_index]
    return node.moves[min_index]


def player_move(node, move):
    global depth, root
    depth += 1
    for i in range(0, len(node.moves)):
        if move == moves:
            root = node.children[i]
            return


moves = [i for i in range(1, 10)]
depth = 0
sum = 0
root = Node()


def introduce_rules():
    print("\t\t----------")
    print("\t\t| SUM 21 |")
    print("\t\t----------")
    print(" Rules:")
    print(" 1. Game will be played between you and the computer: ")
    print(" 2. You can select a number between 1 and 9 without repetition.")
    print(" 3. If the sum of all selected numbers is greater than 21, the last player to select the number loses.")
    print(" 4. If the sum is equal to 21, it will be a tie.")
    print(" 5. If you enter an invalid move, you will be prompted to enter again.")
    print(" Press enter if ready.")


def validate_move(move):
    count = moves.count(move)
    if count > 0:
        return True
    else:
        return False


class _main_:
    move = 0
    introduce_rules()
    input()
    while True:
        try:
            while not validate_move(move) :
                move = int(input("Your move:  "))
            break
        except:
            print("Please enter number only")
    moves.remove(move)
    root.moves.extend(moves)
    minimax(root, depth, move)
    sum += move

    while sum < 21:
        if depth % 2 is not 0:         # Players turn
            print("\t\tCurrent Sum:  " + str(sum))
            print("\t\tAvailable moves: ", moves)
            while True:
                try:
                    while not validate_move(move):
                        move = int(input("Your move:  "))
                    player_move(root, move)
                    break
                except:
                    print("Please enter number only")
        else:                           # AI turn
            move = ai_move(root)
            print(" AI move:  ", move)
        moves.remove(move)
        sum += move
    if sum > 21:
        if depth % 2 is 0:
            print("AI wins!!!")
        else:
            print("You win!!!")
    else:
        print("Its a tie:(")

File: Sum21.java

import java.util.ArrayList;
import java.util.InputMismatchException;
import java.util.Scanner;

/** Date - 20/April/2019
 * @author Tushar Sharma
 */

class Node {
    
    ArrayList<Node> childern;
    ArrayList<Integer> availableMoves;
    ArrayList<Integer> ScorePerMove;
    
    Node() {
        childern = new ArrayList<Node>();
        availableMoves = new ArrayList<Integer>();
        ScorePerMove = new ArrayList<Integer>();
    }
    
}

public class Sum21 {
    
    ArrayList<Integer> currentAvailableMoves;
    int currScore;
    int currDepth;
    Node root;
    
    int aptScore(int depth, Node node) {
        if(depth%2 != 0) {              //AI Mode
            int min = 100;
            for(int i = 0; i < node.ScorePerMove.size(); i++) {
                if(node.ScorePerMove.get(i) < min) {
                    min = node.ScorePerMove.get(i);
                }
            }
            return min;
        } else {                        //Player Mode
            int max = -100;
            for(int i = 0; i < node.ScorePerMove.size(); i++) {
                if(node.ScorePerMove.get(i) > max) {
                    max = node.ScorePerMove.get(i);
                }
            }
            return max;
        }
    }
    
    int calculateScore(int depth, int move) {
        if(move > 21) {
            if(depth%2 != 0) {                  //Player
                return 10 - depth;
            } else {                            //AI
                return depth - 10;
            }
        } else {
            return 0;
        }
    }
    
    boolean checkIfGameOver(int move) {
        return move>=21;
    }
    
    int minimaxTree(Node node, int depth, int move) {
        if(checkIfGameOver(move)) {
            int score =  calculateScore(depth, move);
            node.availableMoves.clear();
            node.ScorePerMove.add(score);
            return score;
        }
        
        int scr = 0;
        
        for(int i = 0; i < node.availableMoves.size(); i++) {
            int currMove = node.availableMoves.get(i);
            node.childern.add(new Node());
            node.childern.get(i).availableMoves.addAll(node.availableMoves);
            node.childern.get(i).availableMoves.remove(i);
            scr = minimaxTree(node.childern.get(i), depth + 1 , move + currMove);
                
            node.ScorePerMove.add(i, aptScore(depth, node.childern.get(i)));
        }
        
        return scr;
    }
    
    int nextMoveAI(Node node) {
        int minIndex = -1;
        int min = 100;
        for(int i = 0; i < node.ScorePerMove.size(); i++) {
            if(node.ScorePerMove.get(i) < min) {
                min = node.ScorePerMove.get(i);
                minIndex = i;
            }
        }
        int move = node.availableMoves.get(minIndex);
        root = node.childern.get(minIndex);
        currDepth++;
        return move;
    }
    
    void nextMovePlayer(Node node, int move) {
        currDepth++;
        root = node.childern.get(node.availableMoves.indexOf(move));
    }
    
    boolean validateMove(int move) {
        for(int i = 0; i < currentAvailableMoves.size(); i++) {
            if(currentAvailableMoves.get(i) == move)
                return true;
        }
        return false;
    }
    
    void introduceRules() {
        System.out.println("\t\t----------");
        System.out.println("\t\t| SUM 21 |");
        System.out.println("\t\t----------");
        System.out.println(" Rules:");
        System.out.println(" 1. Game will be played between you and the computer: ");
        System.out.println(" 2. You can select a number between 1 and 9 without repetition.");
        System.out.println(" 3. If the sum of all selected numbers is greater than 21, the last player to select the number loses.");
        System.out.println(" 4. If the sum is equal to 21, it will be a tie.");
        System.out.println(" 5. If you enter an invalid move, you will be prompted to enter again.");
        System.out.println(" Press enter if ready.");
    }
    
    public static void main(String args[]) {
        Sum21 obj = new Sum21();
        obj.currDepth = 0;
        obj.currentAvailableMoves = new ArrayList<Integer>();
        for(int i = 1; i <= 9; i++) {
            obj.currentAvailableMoves.add(i);
        }
        obj.currScore = 0;
        obj.root = new Node();
        
        Scanner scan = new Scanner(System.in);
        int move = -1;
        
        obj.introduceRules();
        scan.nextLine();
        
        while(true) {
            try{
            do{
                System.out.print("Your move :  ");
                move = scan.nextInt();
            }while(!obj.validateMove(move));
            break;
            } catch(InputMismatchException e) {
                System.out.println("Please enter number only");
                scan.nextLine();
            }
        }
        
        obj.currScore += move;
        
        obj.currentAvailableMoves.remove(obj.currentAvailableMoves.indexOf(move));
        obj.root.availableMoves.addAll(obj.currentAvailableMoves);
        obj.minimaxTree(obj.root, obj.currDepth, move);
        
        while(obj.currScore < 21) {
            
            if(obj.currDepth%2 != 0) {          //Players turn
                System.out.println("\t\tCurrent Sum:  "+ obj.currScore);
                System.out.print("\t\tAvailable moves: ");
                for(int i = 0; i < obj.currentAvailableMoves.size(); i++) {
                    System.out.print(" " + obj.currentAvailableMoves.get(i));
                }
                System.out.println();
                while(true) {
                    try{
                        do{
                            System.out.print("Your move :  ");
                            move = scan.nextInt();
                        }while(!obj.validateMove(move));
                            obj.nextMovePlayer(obj.root, move);
                            break;
                    } catch(InputMismatchException e) {
                        System.out.println("Please enter number only");
                        scan.nextLine();
                    }
                }
            } else {    
                //AI turn
                System.out.print(" AI move:  ");
                move = obj.nextMoveAI(obj.root);
                System.out.println(move);
            }
            obj.currentAvailableMoves.remove(obj.currentAvailableMoves.indexOf(move));
            obj.currScore += move;
        }
        if(obj.currScore > 21) {
            if(obj.currDepth%2 == 0) {
                System.out.println("AI wins!!!");
            } else {
                System.out.println("You win!!!");
            }
        } else {
            System.out.println("Its a tie:(");
        }
    }
    
}
